pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,260 @@
+		for _, cond := range ri.Conditions {
+			if cond.Type == routev1.RouteAdmitted && cond.Status == corev1.ConditionTrue {
+				log.Info("queueing ingresscontroller", "name", ri.RouterName)
+				// Create a reconcile.Request for the router named in the RouteIngress.
+				request := reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      ri.RouterName,
+						Namespace: r.Namespace,
+					},
+				}
+				requests = append(requests, request)
+
+				// Add the Router Name to the currentIngressSet.
+				currentRouteIngresses.Insert(ri.RouterName)
+			}
+		}
+	}
+
+	// Get the previous set of Ingresses of the Route.
+	previousRouteIngresses := r.routeToIngresses[routeNamespacedName]
+
+	// Iterate through the previousRouteIngresses.
+	for routerName := range previousRouteIngresses {
+		// Check if the currentRouteIngresses contains the Router Name. If it does not,
+		// then the Ingress was removed from the Route Status. The reconcile loop is needed
+		// to be run for the corresponding Ingress Controller.
+		if !currentRouteIngresses.Has(routerName) {
+			log.Info("queueing ingresscontroller", "name", routerName)
+			// Create a reconcile.Request for the router named in the RouteIngress.
+			request := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      routerName,
+					Namespace: r.Namespace,
+				},
+			}
+			requests = append(requests, request)
+		}
+	}
+
+	// Map the currentRouteIngresses to Route's NamespacedName.
+	r.routeToIngresses[routeNamespacedName] = currentRouteIngresses
+
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	client    client.Client
+	Namespace string
+	// routeToIngresses stores the Ingress Controllers that have admitted a given route.
+	routeToIngresses map[types.NamespacedName]sets.String
+}
+
+// Reconcile expects request to refer to an Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// If the ingresscontroller is marked to be deleted, then delete the corresponding RouteMetricsControllerRoutesPerShard metric label.
+	if ingressController.DeletionTimestamp != nil {
```

Comment:
I think this should be in the finalizer, not in the Reconcile loop.

Reply:
And I don't understand why you would remove a finalizer?

Reply:
Perhaps this code snippet of how to use finalizers will help to understand why it has to be added and removed in the Reconcile loop: https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#external-resources

Reply:
Moved the deletion of metrics to `ensureIngressDeleted`.